### PR TITLE
refactor: Support all the webs

### PR DESCRIPTION
**Pull request by russell-dot-js**
* Collapse stream-collector-browser and stream-collector-native into stream-collector-web
* Update FetchHttpHandler to return blob if stream not implemented

*Issue #, if available:*
#1107

*Description of changes:*
* Remove stream-collector-browser and stream-collector-native in favor of generic stream-collector-web
* Update FetchHttpHandler to always return blob if stream not implemented

The real changes are [here](https://github.com/aws/aws-sdk-js-v3/pull/1123/files#diff-5a9ed83d70c7c73d9a0b48e9928504d4) and [(starting) here](https://github.com/aws/aws-sdk-js-v3/pull/1123/files#diff-8d3f5954101ca06b03a360d7d88eec5f). Rest is find-replace.

*Open questions*
* Naming of stream-collector-web? I would normally call it stream-collector-client, considering it's for use client-side, but don't want it to be easily confused with the actual (AWS) clients in the sdk. Could also go for something like stream-collector-fetch to demonstrate coupling with FetchHttpHandler
* Ok that since, this is technically a new package, I started it at 1.0.0-beta.0?
* What if we collapsed the "HttpHandlers" and "StreamCollectors" into a single package? E.g. rather than mixing and matching HttpHandlers and StreamCollectors, they always come together. So if you use FetchHttpHandler, you get the FetchStreamCollector. etc.
* Can we get rid of the "bufferBody" option in FetchHttpHandler? There's no reason for native to say "never give me a stream" - my assumption is that it will be supported, eventually, and as long as we support both scenarios at run-time, we're good to go.
* Is there a preference to keep stream-collector-browser and stream-collector-native separate, even though they'd both delegate to this same package? My thought was that since this is still in beta, and we can get away with deleting stuff, I didn't see a reason to create (what's arguably) technical debt.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by russell-dot-js** [comment]
The real changes are [here](https://github.com/aws/aws-sdk-js-v3/pull/1123/files#diff-5a9ed83d70c7c73d9a0b48e9928504d4) and [(starting) here](https://github.com/aws/aws-sdk-js-v3/pull/1123/files#diff-8d3f5954101ca06b03a360d7d88eec5f). Rest is find-replace.

**Comment by russell-dot-js** [comment]
Crap. I forgot to update stream-collector-web/CHANGELOG after the git mv from stream-collector-native. Will wait for feedback to open questions in PR description before buttoning up.

File: packages/fetch-http-handler/src/fetch-http-handler.ts [hunk 1]
```diff
@@ -21,6 +21,10 @@ export interface BrowserHttpOptions {
    * ReadableStream.
    * This option is only useful in ReactNative.
    */
+  // TODO can we get rid of this? is there a scenario where a client wants to
+  //  insist on a Blob, even if stream is implemented? regardless, comment above
+  //  is outdated, this is not only useful in ReactNative due to variety of fetch
+  //  implementations
```
Comment (AllanFly120) [inline packages/fetch-http-handler/src/fetch-http-handler.ts:27]: I think the same way. If we rely on dynamicly detecting the `ReadableStream` implementation in the environment, we should remove it.

**Comment by AllanFly120** [comment]
Hey @russell-dot-js 

Very awesome PR! Regarding your questions:

> Is there a preference to keep stream-collector-browser and stream-collector-native separate, even though they'd both delegate to this same package? My thought was that since this is still in beta, and we can get away with deleting stuff, I didn't see a reason to create (what's arguably) technical debt.

Given we prefer to return body in `blob` or `ReadableStream` dynamicly from `fetch-http-handler`, we should also merge the 2 difference stream collector. 

> What if we collapsed the "HttpHandlers" and "StreamCollectors" into a single package? E.g. rather than mixing and matching HttpHandlers and StreamCollectors, they always come together. So if you use FetchHttpHandler, you get the FetchStreamCollector. etc.

Unfortunately we cannot do it. For any service/operations supporting streaming payload, we don't call StreamCollectors to the response, instead we return the stream from low-level http handler to users directly. In these case, the http handlers doesn't have the prior knowledge of whether the response is from an operation with streaming member. 

Generally very solid PR. I'd like to get some more set of eyes the review the solution. Ones we decide to go with this solution, I can provide some help with the code generation. Basically, we don't work directly on `clients` folder, because they are generated. Instead, we make code change to our [code generator in Java](https://github.com/aws/aws-sdk-js-v3/tree/master/codegen), or the lower level [smithy typescript generator](https://github.com/awslabs/smithy-typescript). Don't worry about it just now, we will provide all the guides needed.

**Comment by russell-dot-js** [comment]
> Hey @russell-dot-js
> 
> Very awesome PR! 

Thank you!! Happy to help.

> > Is there a preference to keep stream-collector-browser and stream-collector-native separate, even though they'd both delegate to this same package? My thought was that since this is still in beta, and we can get away with deleting stuff, I didn't see a reason to create (what's arguably) technical debt.
> 
> Given we prefer to return body in `blob` or `ReadableStream` dynamicly from `fetch-http-handler`, we should also merge the 2 difference stream collector.

Noted

> > What if we collapsed the "HttpHandlers" and "StreamCollectors" into a single package? E.g. rather than mixing and matching HttpHandlers and StreamCollectors, they always come together. So if you use FetchHttpHandler, you get the FetchStreamCollector. etc.
> 
> Unfortunately we cannot do it. For any service/operations supporting streaming payload, we don't call StreamCollectors to the response, instead we return the stream from low-level http handler to users directly. In these case, the http handlers doesn't have the prior knowledge of whether the response is from an operation with streaming member.

Curios -- are you sure we can't do this?

Example:
```
interface HttpProvider<T> {
  handler: HttpHandler<T>;
  streamCollector: StreamCollector<T>;
}

// could be generic or not, doesn't really matter - ultimately here T might be ReadableStream | Blob for web, or Readable for node.
```
For the S3 client, the streamCollector would be ignored, but for most clients, it would be used.

> Generally very solid PR. I'd like to get some more set of eyes the review the solution. Ones we decide to go with this solution, I can provide some help with the code generation. Basically, we don't work directly on `clients` folder, because they are generated. Instead, we make code change to our [code generator in Java](https://github.com/aws/aws-sdk-js-v3/tree/master/codegen), or the lower level [smithy typescript generator](https://github.com/awslabs/smithy-typescript). Don't worry about it just now, we will provide all the guides needed.

Got it! I knew there was some code gen for the clients, but assumed that since the clients were checked in, they were not code-genned. For my own understanding, can you please clarify why the generated code is checked in? Is this to speed up CI, to keep every change in version control to ease human debugging, or to facilitate one-off edits of generated code (seems risky)?

Also, I am curious why I have tests failing in CI that did not pass locally. `yarn test:all` passed locally... perhaps I should open an issue regarding unclear documentation around how to test? I'm also happy to look into #1092, adding more transparency to the CI process would help "outsiders" like me contribute.

**Comment by russell-dot-js** [comment]
@AllanFly120 I've amended my commit (do you prefer squashed PR's?) removing "bufferBody", and fixing the failing tests. I had to update the tests to not return "body", as the tests were returning strings, not ReadableStreams, and we do not have a ReadableStream polyfill set up for the node-js test environment (do we want one? It would be nice to add some unit tests demonstrating that it properly handles ReadableStream if implemented, and falls back to blob.)

See changes in packages/fetch-http-handler/src/fetch-http-handler.ts and packages/fetch-http-handler/src/fetch-http-handler.spec.ts

**Comment by russell-dot-js** [comment]
@AllanFly120 / @trivikr just wanted to check in on this? Per https://github.com/aws/aws-sdk-js-v3/pull/1121#issuecomment-623937542, @stereobit also has users in a tough spot right now. Maybe we can at least release the quick fix in #1121 while this is reviewed?

**Comment by AllanFly120** [comment]
@russell-dot-js Hi, sorry for the late response. Can you remove all the diffs in `clients` folder from this PR and only keep the `fetch-http-handler` and `event-collector-web` changes? I will open another PR with updated clients to coordinate with this PR.

File: packages/fetch-http-handler/src/fetch-http-handler.ts [hunk 2]
```diff
@@ -80,8 +72,16 @@ export class FetchHttpHandler implements HttpHandler {
           transformedHeaders[pair[0]] = pair[1];
         }
 
+        // todo is this safe? what if different polyfills / libraries conflict
+        //   with eachother, e.g. a polyfilled response.body in a browser with
+        //   native window.ReadableStream.
+        //   can we just check for existence of response.body, as that should be
+        //   guaranteed to be a ReadableStream if it exists?
+        const hasResponseStream = typeof ReadableStream === 'function' &&
+          response.body instanceof ReadableStream;
```
[the remark this answers is not in the document]
Comment (russell-dot-js) [reply packages/fetch-http-handler/src/fetch-http-handler.ts]: Done and done

Was hoping you could response to my questions here:
https://github.com/aws/aws-sdk-js-v3/pull/1123#issuecomment-621454314
and here?
https://github.com/aws/aws-sdk-js-v3/pull/1123#issuecomment-621469941

**Comment by AllanFly120** [comment]
Hi @russell-dot-js 
> Curious -- are you sure we can't do this?

After putting more thoughts on it I think:

✅ we can colapse `fetch-request-handler` and `stream-collector-web` into the same package; Also `node-http-handler` and `stream-collector-node` into the same package. Because they always co-exist.
❌ we should not put the handler and stream collector under the same interface(e.g. `nodeHttpProvider`). Because they belong to different request handling stage and should be decoupled for extensibility of SDK, here's details:

Request handler are supposed to be a thin layer over the runtime primitives, it must return the raw low level response payload. It serves as [the core of service client's middleware stack](https://github.com/aws/aws-sdk-js-v3/blob/5f820e6a06de21965db1ac4094b33b04d90533a3/clients/client-accessanalyzer/commands/CreateAnalyzerCommand.ts#L60-L64) regardless of the service model. However, stream collector is a util function, although very commonly used, that defined by the service model. It is [generated when the service needs structured response](https://github.com/aws/aws-sdk-js-v3/blob/5f820e6a06de21965db1ac4094b33b04d90533a3/clients/client-accessanalyzer/protocols/Aws_restJson1_1.ts#L969). 

Why is this difference important? Because V3 SDK is designed with a lot of flexibility. Any AWS client is extended from [a white-labeled client](https://github.com/aws/aws-sdk-js-v3/blob/5f820e6a06de21965db1ac4094b33b04d90533a3/packages/smithy-client/src/client.ts#L9-L30) that only knows the low-level request handler(`requestHandler`). This design enables us to support any non-http services in the future like WebSocket, [MQTT](http://mqtt.org/) and even unstandard JS runtime like [JerryScript](https://jerryscript.net/). In the future, we'd like to support everyone to generate their own V3-like SDK on top of the white-labeled client, it doesn't make much sense if we require a stream collector in the white-labeled client. 

I hope my explanation makes sense to you.

**Comment by AllanFly120** [comment]
@russell-dot-js 
> It would be nice to add some unit tests demonstrating that it properly handles ReadableStream if implemented, and falls back to blob.)

We used string body because we simply run the test in Node runtime. It will be awesome if we ca have a webstream polyfill test.

I think you can add the test in a separate PR if you want because many Amplify users are still blocked by this issue in SDK. I'd like get this fix merged in earlier

> Were those scopes accidentally committed, or is the intent of test:all to only test those two packages?

The scoped test script only runs in the 2 packages. Because the unit test in the packages are run in puppeteer, so we call the test script in the 2 packages explicitly.

**Comment by AllanFly120** [comment]
@russell-dot-js 
Hi, I'm working on another PR adopting your commits from this PR. I will post the PR when it's ready.

**Pull request closed without merging**